experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp

```diff
@@ -1291,6 +1291,172 @@ static Value buildStateCreation(ValueToStreamOp op,
+        upstreamStateType, updatedState, b.getIndexAttr(index));
+
+    // Call Open on upstream.
+    SymbolRefAttr openFunc = upstreamInfo.openFunc;
+    auto openCallOp = b.create<func::CallOp>(openFunc, upstreamStateType,
+                                             initialUpstreamState);
+
+    // Update state.
+    Value updatedUpstreamState = openCallOp->getResult(0);
+    updatedState = b.create<iterators::InsertValueOp>(
+        updatedState, b.getIndexAttr(index), updatedUpstreamState);
+  }
+
+  return updatedState;
+}
+
+/// Builds IR that calls next on all upstream iterators and assembles an output
+/// element from the resulting elements. Pseudo-code:
+///
+/// nextElement = tuple()
+/// for each upstream in upstreams:
+///   if nextUpstreamElement = upstream->Next():
+///     nextElement.append(nextUpstreamElement)
+///   else:
+///   return {}
+/// return nextElement
+///
+/// Possible output (for one input stream):
+///
+/// %1 = iterators.extractvalue %initialState[0] :
+///          !iterators.state<!upstream_state_type>
+/// %2:3 = call @iterators.map.next.0(%1) :
+///            (!upstream_state_type) -> (!upstream_state_type, i1, i32)
+/// %3 = arith.andi %true, %2#1 : i1
+/// %state = iterators.insertvalue %2#0 into %initialState[0] :
+///              !iterators.state<!upstream_state_type>
+/// %tuple = tuple.from_elements %2#2 : tuple<i32>
+/// return %state, %3, %4 :
+///     !iterators.state<!!upstream_state_type>, i1, tuple<i32>
+static llvm::SmallVector<Value, 4>
+buildNextBody(ZipOp op, OpBuilder &builder, Value initialState,
+              ArrayRef<IteratorInfo> upstreamInfos, Type elementType) {
+  Location loc = op.getLoc();
+  ImplicitLocOpBuilder b(loc, builder);
+  Type i1 = b.getI1Type();
+
+  // Call next on each upstream.
+  Value updatedState = initialState;
+  Value hasNext = b.create<arith::ConstantIntOp>(/*value=*/1, /*width=*/1);
+  SmallVector<Value> upstreamElements;
+  for (auto [index, upstreamInfo] : llvm::enumerate(upstreamInfos)) {
+    Type upstreamStateType = upstreamInfo.stateType;
+    auto inputStreamType = op->getOperand(index).getType().cast<StreamType>();
+    Type inputElementType = inputStreamType.getElementType();
+
+    // Extract upstream state.
+    Value initialUpstreamState = b.create<iterators::ExtractValueOp>(
+        upstreamStateType, updatedState, b.getIndexAttr(index));
+
+    // Call next on upstream.
+    SmallVector<Type> nextResultTypes = {upstreamStateType, i1,
+                                         inputElementType};
+    SymbolRefAttr nextFunc = upstreamInfo.nextFunc;
+    auto nextCall =
+        b.create<func::CallOp>(nextFunc, nextResultTypes, initialUpstreamState);
+    Value upstreamHasNext = nextCall->getResult(1);
+    Value upstreamElement = nextCall->getResult(2);
+
+    // Combine hasNext value of the call with previous ones.
+    hasNext = b.create<arith::AndIOp>(hasNext, upstreamHasNext);
```

Comment:
We need more description for the termination state.
You could also envision having attributes so your op works with zip_min/zip_equal and other types of semantics, asserting properly.

Reply:
OK, documented in TD file and added as TODO.